feat: initial implementation

lentidas opened the pull request:
## Description of the changes

This is the first implementation of the DevOps Stack module used to provision an SKS cluster on Exoscale.

It is a port and merge of the resources deployed by our [legacy DevOps Stack module](https://github.com/camptocamp/devops-stack/tree/v0-monolithic/modules/sks/exoscale) as well as our in-house [SKS module](https://github.com/camptocamp/terraform-exoscale-sks/tree/main).

A major refactor and cleanup was done from the code existing in these modules. The module now has better descriptions for all outputs and variables.

**The main difference on the code behavior is the way the router node pool and load balancer are created.**
Instead of requiring the user to pass the configuration themselves, it is now hardcoded and then merged with the remaining node pools that are required to be passed as a variable. The load balancer is created by default inside the module, but without any service. Now, they will be created by the Traefik module, which is the reason that this module outputs multiple IDs related to the router node pool and load balancer.

The PR on the Traefik module related to this one -> https://github.com/camptocamp/devops-stack-module-traefik/pull/44

Other notable changes:
- validation of the node pool input
- use the `exoscale_sks_kubeconfig` resource
- use Cilium as the CNI by default but give the user the option to choose otherwise

lentidas committed:
style: move security rules to their own file

lentidas committed:
ci: point existing workflows to main branch and add project workflow

lentidas committed:
feat: add first working code

lentidas committed:
chore: cleanup code and comments

lentidas committed:
docs(terraform-docs): generate docs and write to README.adoc

ckaenzig commented on variables.tf:
```diff
@@ -4,29 +4,52 @@ variable "cluster_name" {
 
 variable "base_domain" {
-  description = "The base domain used for Ingresses."
+  description = "The base domain used for Ingresses. If not provided, nip.io will be used taking the NLB IP address."
```
I'm not sure it's a good idea to make this variable optional. It makes sense to use nip.io for deployments on KinD for lcoal tests. But IMO we should mandate having an actual domain for all stacks deployed on hosted platforms.

lentidas replied:
I agree that maybe creating the domain itself inside the module could be a bad idea, I did not consider that.

As for the use of nip.io, we already did the same on the EKS cluster, and the DNS record is not created unless you pass the `base_domain` variable (the domain is still created outside, as far as I know).

ckaenzig replied:
Ok, then this would need to be discussed more globally for all cluster modules. So for now it's fine to make this config optional to be consistent across clusters.

ckaenzig commented on dns.tf:
```diff
@@ -1,9 +1,23 @@
-data "exoscale_domain" "this" {
-  name  = local.base_domain
+resource "exoscale_domain" "this" {
```
If a domain is used, I don't think this module should create it, because you most likely won't want to delete it if the stack is destroyed. And it should be in line with the behavior of other cluster modules in that regard.

ckaenzig commented on variables.tf:
```diff
@@ -4,29 +4,52 @@ variable "cluster_name" {
 
-variable "zone" {
-  description = "The name of the zone to deploy the SKS cluster into."
+variable "cluster_version" {
```
In the EKS and KinD modules, this variable is named `kubernetes_version` which is more meaningful in my opinion. I think it would be better to use the same name here.

ckaenzig commented on variables.tf:
```diff
@@ -4,29 +4,52 @@ variable "cluster_name" {
 
-variable "kubernetes_version" {
-  description = "Kubernetes version to use for the SKS cluster."
+variable "cluster_zone" {
```
The main point of this module is to deploy an SKS cluster, so I don't think the `cluster_` prefix in variables `cluster_zone` or `cluster_auto_upgrade` is very meaningful. We could same the same for `cluster_name`, but that's a variable we're already using on other modules.

ckaenzig reviewed (commented):
Some comments to be discussed, otherwise I don't have any big issue with this PR.

lentidas committed:
refactor: change variable names

lentidas committed:
docs(terraform-docs): generate docs and write to README.adoc

lentidas committed:
fix: remove domain creation from inside the module

lentidas committed:
docs(terraform-docs): generate docs and write to README.adoc

lentidas committed:
fix: remove data source for the domain

The apply fails even when the var.base_domain depends on a resource only known on apply, this is because Terraform infers automatically the domain name even before creating the Exoscale domain on the caller module.

lentidas committed:
docs(terraform-docs): generate docs and write to README.adoc

ckaenzig commented on lb_dns.tf:
```diff
@@ -0,0 +1,14 @@
+  name = format("ingress-%s", var.cluster_name)
+}
+
+resource "exoscale_domain_record" "wildcard_colorized" {
```
I wouldn't call this "colorized", because it's just a DNS record that includes the name of the cluster, whether it's a color or not. Also it's the only wildcard record this module manages, so no need to distinguish it.

ckaenzig commented on variables.tf:
```diff
@@ -9,6 +9,12 @@ variable "base_domain" {
   default     = null
 }
 
+variable "domain_id" {
```
In the EKS module, we don't have this variable but we use a datasource to find the domain id from the `base_domain` variable. I guess the reason is that the domain is often not managed in the same terraform state as the devops-stack. I'm not sure if it's better to do the same in this module or if there's a reason to do it differently... :shrug: ?

lentidas replied:
I think we can do it similarly, it's just that I probably had some issues with it in the beggining and then never came back to it. I'll add a data source and test it.

lentidas commented:
I'll turn this PR into a draft again to further continue testing and because it could possibly change until we have a fully working deployment working.

I would still like to test:
- [x] review the way we use the kubeconfig resource, because as it stands, if the terraform state is not ran in the interval available to update the kubeconfig, Terraform will throw an error saying it cannot connect to the cluster.
- [x] the domain creation: I found out last week that I did not get it to work properly, but with nip.io it works;
- [x] the cluster auto-upgrade: I added it from the beginning but I've yet to test it;
- [x] consider using Cilium as a CNI instead of Calico, since it seems to be easily supported by SKS clusters, although this can be an improvement down the line instead of including it in this initial implementation;

lentidas committed:
fix: adjust the TTL of the generated kubeconfig

lentidas committed:
feat: add Cilium as a CNI option

lentidas committed:
docs(terraform-docs): generate docs and write to README.adoc

lentidas committed:
feat: change CNI variable so it's easier to change default

lentidas committed:
docs(terraform-docs): generate docs and write to README.adoc

lentidas committed:
fix: remove the need to have the domain_id variable

lentidas committed:
docs(terraform-docs): generate docs and write to README.adoc

lentidas committed:
fix: add CoreDNS ConfigMap to solve issue with DNS autodiscovery

lentidas committed:
docs(terraform-docs): generate docs and write to README.adoc

lentidas committed:
fix: remove CoreDNS config overload

lentidas committed:
fix: add cluster ID variable to solve DNS issue

lentidas committed:
docs(terraform-docs): generate docs and write to README.adoc

hbollon committed:
feat: parameterize default router nodepool (#4)

* feat: parameterize default router nodepool

* docs(terraform-docs): generate docs and write to README.adoc

---------

Co-authored-by: hbollon <hbollon@users.noreply.github.com>

hbollon committed:
docs: manual k8s upgrade strategy documentation (#3)

lentidas committed:
chore: terraform fmt

lentidas committed:
feat: improve structure validation of the nodepool variables

lentidas committed:
feat: improvements to the kubeconfig resources

lentidas committed:
docs: continue docs even further

lentidas committed:
docs(terraform-docs): generate docs and write to README.adoc

hbollon committed:
docs: explain Longhorn manipulations during k8s upgrade (#5)

lentidas committed:
chore: remove last TODOs

lentidas committed:
fix: set kubeconfig TTL at 0 by default

lentidas committed:
docs(terraform-docs): generate docs and write to README.adoc

ckaenzig commented on variables.tf:
```diff
@@ -4,29 +4,104 @@ variable "cluster_name" {
+}
+
+variable "cni" {
+  description = "Specify which CNI to use by default. Accepted values are `calico` or `cilium`, but you cannot change this value after the first deployment. This module creates the required security group rules."
```
```suggestion
  description = "Specify which CNI plugin to use (cannot be changed after the first deployment). Accepted values are `calico` or `cilium`. This module creates the required security group rules."
```

ckaenzig commented on variables.tf:
```diff
@@ -4,29 +4,104 @@ variable "cluster_name" {
-variable "zone" {
-  description = "The name of the zone to deploy the SKS cluster into."
+variable "auto_upgrade" {
+  description = "Enable automatic upgrade of the SKS cluster."
```
```suggestion
  description = "Enable automatic upgrade of the SKS cluster control plane."
```

ckaenzig reviewed (dismissed):
Suggested some small documentation changes, otherwise it looks good.

lentidas committed:
docs: apply suggestions from code review

Co-authored-by: Christian Kaenzig <christian.kaenzig@camptocamp.com>

lentidas committed:
docs(terraform-docs): generate docs and write to README.adoc
